networkx/algorithms/assortativity/mixing.py

```diff
@@ -188,8 +191,11 @@ def numeric_mixing_matrix(G, attribute, nodes=None, normalized=True):
     s = set(d.keys())
     for k, v in d.items():
         s.update(v.keys())
-    m = max(s)
-    mapping = {x: x for x in range(m + 1)}
+    if compressed:
+        mapping = {x: i for x, i in zip(s, range(len(s)))}
+    else:
+        m = max(s)
+        mapping = {x: x for x in range(m + 1)}
```

Comment:
Would the following code do what this PR is proposing? 
Can you describe why someone would want the compression option and other people not want is compressed?
I don't understand why someone would want the empty rows and columns.
```
    # collect all possible values of the attribute
    s = set(d)
    s.update(attr_value for key, nbrs in d.items() for att_value in nbrs)
    if compressed:
        # include only attribute values that appear in G for some node
        mapping = {x: i for i, x in enumerate(s)}
    else:
        # include each integer even if unused in G 
        mapping = {x: x for x in range(max(s) + 1)}
```

Reply:
Compression parameter is added for compatibility: if I want to calculate a mixing matrix, then I would need these empty rows and columns.
In order to calculate assortativity coefficient we need a mixing matrix, and _I assumed_ that deleting empty rows and columns would significantly help with calculation time and still give correct answers. 
It really helped with time, but answers were wrong, and as for now I haven't yet found the cause of the error

Reply:
Reading the papers by Newman, it seems that the difference these "empty rows" make is that they aren't empty. They are filled with zeros. So, if you average or find standard deviation those values could be quite different if you remove the zeros. 

There may be a way to do the computation without storing all the zeros. I haven't worked through it, but if we could use scipy.sparse matrices instead of arrays -- or maybe just work with the dict-of-dict which is also a sparse matrix representation.  We'd have to compute the sigmas and averages using formulas instead of the traditional functions. But it seems reasonable that we could make it fast and also correct. Certainly we don't need to store all those zeros to get the answer.

Reply:
Yeah, my approach was wrong.
I believe it is better for me to close this pull-request and make another one when my solution is ready — the description and even name for this PR are kind of wrong